src/librustc_trans/debuginfo/metadata.rs

```diff
@@ -669,7 +678,10 @@ pub fn file_metadata(cx: &CrateContext, path: &str, full_path: &Option<String>)
             }
         });
 
-    file_metadata_(cx, path, file_name, &work_dir)
+    let sess = cx.sess();
+    let remap_file_name = remap_path(sess, file_name);
```

Comment:
This ends up matching against a full path rather than the path passed on the command line. The `let file_name` expression above ends up using the `full_path` parameter if its present (which it always appears to be) rather than the relative `path` (which it goes to some lengths to make relative to `work_dir`). I'm not sure what the larger intent is here, but the effect is that this doesn't work as expected. 

It looks like this was done deliberately in 24e74916 by @luser. This suggests that path remapping should be done in absolute space - though the code is pretty inconsistent. If `full_path` is `None`, then I think the resulting path will be relative.